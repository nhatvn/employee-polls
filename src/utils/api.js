import {
  _getQuestions,
  _getUsers,
  _saveQuestion,
  _saveQuestionAnswer
} from './_DATA.js'

export function getInitialData() {
  return _getUsers().then((users) => ({
    users,
  }))
}

export function getInitialDataAfterLogin() {
  return _getQuestions().then((questions) => ({
    questions,
  }))
}

export function saveQuestion(question) {
  return _saveQuestion(question);
}

export function saveQuestionAnswer(answer) {
  return _saveQuestionAnswer(answer);
}