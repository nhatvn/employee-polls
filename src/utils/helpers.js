import { useLocation, useNavigate, useParams } from "react-router-dom";

export function formatDate(timestamp) {
  if (timestamp === null) {
    return '';
  }

  const d = new Date(timestamp);
  const time = d.toLocaleTimeString('en-US');
  return time.substr(0, 5) + time.slice(-2) + ' | ' + d.toLocaleDateString();
}

export function withRouter(Component) {
  const ComponentWithRouterProp = (props) => {
    let location = useLocation();
    let navigate = useNavigate();
    let params = useParams();
    return <Component {...props} router={{ location, navigate, params }} />;
  };

  return ComponentWithRouterProp;
};

export function getImages(image) {
  if (image === null) {
    return null;
  }
  else {
    return require.context('../images', true)(`./${image}`);
  }
}

export function objectToArray(objectInput) {
  let list = [];

  for (const key in objectInput) {
    if (Object.hasOwnProperty.call(objectInput, key)) {
      const element = objectInput[key];
      list.push(element);
    }
  }

  return list;
}

export function formatLeaderboard(users, questions) {
  let leaderboardList = [...objectToArray(users)];
  let questionList = objectToArray(questions);
  leaderboardList.forEach((leaderboard) => {
    leaderboard.answered = 0;
    leaderboard.created = questionList.filter((question) => {
      leaderboard.answered += question.optionOne.votes.filter((vote) => {
        return vote === leaderboard.id;
      }).length;

      leaderboard.answered += question.optionTwo.votes.filter((vote) => {
        return vote === leaderboard.id;
      }).length;

      return question.author === leaderboard.id;
    }).length;
  });

  return leaderboardList;
}