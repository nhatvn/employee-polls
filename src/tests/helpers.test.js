import { _getQuestions, _getUsers } from "../utils/_DATA";
import { formatDate, formatLeaderboard, objectToArray } from '../utils/helpers'

describe('objectToArray', () => {
    it('input object will return the array', () => {
        var result = objectToArray({});

        expect(result).toEqual([]);
    })
})

describe('formatDate', () => {
    it('input object will return the array', () => {
        var result = formatDate(1489579767190);

        expect(result).toEqual('7:09:PM | 3/15/2017');
    })

    it('input object will return null', () => {
        var result = formatDate(null);

        expect(result).toEqual('');
    })
})

describe('formatLeaderboard', () => {
    it('will match snapshot', async () => {
        const users = await _getUsers();
        const questions = await _getQuestions();

        var result = formatLeaderboard(users, questions);

        expect(result).toMatchSnapshot();
    })
})