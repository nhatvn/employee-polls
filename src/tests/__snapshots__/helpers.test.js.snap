// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatLeaderboard will match snapshot 1`] = `
Array [
  Object {
    "answered": 4,
    "answers": Object {
      "6ni6ok3ym7mf1p33lnez": "optionOne",
      "8xf0y6ziyjabvozdd253nd": "optionOne",
      "am8ehyc8byjqgar0jgpub9": "optionTwo",
      "loxhs1bqm25b708cmbf3g": "optionTwo",
    },
    "avatarURL": "woman-user-color-icon.png",
    "created": 2,
    "id": "sarahedo",
    "name": "Sarah Edo",
    "password": "password123",
    "questions": Array [
      "8xf0y6ziyjabvozdd253nd",
      "am8ehyc8byjqgar0jgpub9",
    ],
  },
  Object {
    "answered": 2,
    "answers": Object {
      "vthrdm985a262al8qx3do": "optionOne",
      "xj352vofupe1dqz9emx13r": "optionTwo",
    },
    "avatarURL": "man-user-color-icon.png",
    "created": 2,
    "id": "tylermcginnis",
    "name": "Tyler McGinnis",
    "password": "abc321",
    "questions": Array [
      "loxhs1bqm25b708cmbf3g",
      "vthrdm985a262al8qx3do",
    ],
  },
  Object {
    "answered": 3,
    "answers": Object {
      "6ni6ok3ym7mf1p33lnez": "optionOne",
      "vthrdm985a262al8qx3do": "optionTwo",
      "xj352vofupe1dqz9emx13r": "optionOne",
    },
    "avatarURL": "boy-glasses-icon.png",
    "created": 2,
    "id": "mtsamis",
    "name": "Mike Tsamis",
    "password": "xyz123",
    "questions": Array [
      "6ni6ok3ym7mf1p33lnez",
      "xj352vofupe1dqz9emx13r",
    ],
  },
  Object {
    "answered": 1,
    "answers": Object {
      "xj352vofupe1dqz9emx13r": "optionOne",
    },
    "avatarURL": "elderly-grandmother-icon.png",
    "created": 0,
    "id": "zoshikanlu",
    "name": "Zenobia Oshikanlu",
    "password": "pass246",
    "questions": Array [],
  },
]
`;
